Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -1315,6 +1317,64 @@ def visit_ExprNode(self, node):
         # With statements are never inside expressions.
         return node
 
+class ComprehensionScopeTransform(CythonTransform, SkipDeclarations):
+    """
+    Moves expressions out of list/set/etc comprehensions so they have the correct scope
+
+    Effectively
+        cmp = [ a for a in some_expression ]
+    is changed to
+        tmp = some_expression
+        cmp = [ a for a in tmp ]
+
+    (with some complications to keep optimizations (e.g. if "some_expression" is a range
+    where possible)
+    """
+    def visit_ComprehensionNode(self, node):
+        itseq = node.loop.iterator.sequence
+        # there's no value in substituting these very simple expressions
+        # with a temp (but NameNode must be so that it's evaluated in
+        # the right scope)
+        if isinstance(itseq, ExprNodes.ConstNode):
+            return node
+        original_node = node
+
+        new_itseq = ComprehensionResultRefNode(itseq)
+        node.loop.iterator.sequence = new_itseq
+        node = ComprehensionEvalWithTempExprNode(new_itseq, node)
+
+        if (isinstance(itseq, ExprNodes.SimpleCallNode) and
+            len(itseq.args) <= 3 # range with 3 arguments is the
+                  # most complicated expression that may be optimizable
```

Comment:
"complex". Three arguments don't seem "complicated" to me, at least.

Reply:
It's only worth creating these additional substitutions for the cases that are optimized when the original node is replaced and so the inputs to it need evaluating in the correct scope/at the correct time.

That's the various `dict`/`set` iterations (most complicated is: `expr.single_arg_call()`), C arrays (not a call anyway), `enumerate` (`enumerate` is at most 2 args, although `start` might be a keyword argument so I should go back and check I haven't broken anything there), `reversed` (one argument), or `range` (1-3 arguments).

Since nothing more "complicated" can currently be substituted during the optimization that's why it stops there. Maybe it should be removed as being a pointless check because it doesn't do any harm (maybe a tiny bit of speed?) to process more complicated calls.